llvm/test/tools/llvm-cxxfilt/demangle-constrained-template.test

```diff
@@ -0,0 +1,3 @@
+RUN: llvm-cxxfilt -n _ZN3FooIiE6methodITk4TrueIT_EiEEvS3_ | FileCheck %s
+
+CHECK: void Foo<int>::method<int>(T)
```

Comment:
This looks weird to me actually. I think a user would be confused as to where this `T` here came from. I don't think we ever output the generic template parameter names here. Probably what happened here is that the `S3_` substitution picks up the fake constraint argument name that we deduced because of ` HasIncompleteTemplateParameterTracking`.

I think what we really want to do is track the parameter levels properly, so that constraint expressions have access to template parameters in enclosing scopes. Currently `parseTemplateArgs` creates a fresh template parameter level because (from what I can tell) the demangler assumes we never reference template parameters from outer levels. But that's not the case anymore with constraints it looks like. Maybe just avoiding clearing the level if we're parsing a `Tk` is enough? I'm not sure atm, but maybe @zygoloid @urnathan have some better ideas here

Reply:
> I  don't think we ever print out generic template parameter names

I think @VitaNuo showed me an example where we do already print those names (in constraints, not in lambda parameters). Hence, this workaround made sense until the proper fix lands.
@VitaNuo do you have those examples?

> Maybe just avoiding clearing the level if we're parsing a Tk is enough? 

It's probably more complicated than this. Apart from `Tk`, we should also handle constraints (`Q` IIRC). And the actual parameters we track do need to be reset at certain points when the new "top-level" name is being parsed. To put it another way, the parameter numbering is not "global" to the mangled name, instead the same numbering can mean different types at different points of the mangled name and we are not sure what are the points at which the template numbering should be reset or backtracked. @VitaNuo is in the process of figuring that out.

Also wanted to share https://gcc.godbolt.org/z/dqT877q6K and https://gcc.godbolt.org/z/bcfWqGdGz that we came up with yesterday that use `TL<n>` numbering referencing different types at different points in the mangled name.

PS Maybe the solution you have in mind already covers that, but I wasn't sure, so wanted to bring this up.

Reply:
> It's probably more complicated than this. Apart from Tk, we should also handle constraints (Q IIRC). And the actual parameters we track do need to be reset at certain points when the new "top-level" name is being parsed. To put it another way, the parameter numbering is not "global" to the mangled name, instead the same numbering can mean different types at different points of the mangled name and we are not sure what are the points at which the template numbering should be reset or backtracked.

Yea I imagined this to be more complicated than I made it out to be. Happy to land it as is since it's better than failing to demangle at all. And there's precedent for printing the template numbering with constraint expressions.